WordPress/build.gradle

```diff
@@ -270,6 +270,22 @@ android {
         checkDependencies = true
         checkGeneratedSources = true
         lintConfig file("${project.rootDir}/config/lint.xml")
+
+        /*
+        There are several bugs reported for several lifecycle-livedata versions. Here is a couple:
+        * https://issuetracker.google.com/issues/169249668
+        * https://issuetracker.google.com/issues/189967522
+
+        There are several bug fix reports in ther lifecycle release notes. I can see 5 bug fixes
+        related to it if I search for `NullSafeMutableLiveData`:
+        * https://developer.android.com/jetpack/androidx/releases/lifecycle NullSafeMutableLiveData)
+
+        We are disabling this check for now as it results in a crash, but we should try to enable
+        it again as we update our `lifecycle` version.
+
+        Disabling this from `lint.xml` file doesn't seem to work, that's why it's added here.
+        */
+        disable "NullSafeMutableLiveData"
```

Comment:
Suggestion (💡): Having said [the above](https://github.com/wordpress-mobile/WordPress-Android/pull/16632/files#r881611141) I still believe that this `NullSafeMutableLiveData ` check is a valid check, that produces good results, even with our current `2.2.0` version of `lifecycle`. As such, I wouldn't recommend disabling it at this point.

Out of the 35, I can totally recommend suppressing 10 of them (inline and closer to the source), as they indeed seem false positives to me, but for the rest I recommend fixing them. I actually spend a bit of time on it and the fixes are not that hard. I can share the patch I ended up with if you like.

PS: Maybe the newest `2.4.1` version of the `lifecycle` library would automatically fix those 10 false positives that I noticed. 🤷

Reply:
@ParaskP7 If we don't disable the check, lint task crashes due to a bug in `NullSafeMutableLiveData` checks. This is explained in [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/16632/files#diff-276cc53d974bad3212f8e1f709d32ea3f8d380c4dccb10419505d283fb41f2d7R283-R284):

> We are disabling this check for now as it results in a **crash**, but we should try to enable it again as we update our `lifecycle` version.

Reply:
@oguzkocer This is very interesting that link task crashed on your side, it doesn't crash for me locally, do you mean it crashed on CI maybe? 🤔 Locally for me it completes successfully with a failure because of those 35 new errors. Maybe I am still misunderstanding something in terms of the task crashing... 🤷

Reply:
PS: After removing `disable "NullSafeMutableLiveData"`, I actually managed to run Lint successfully after fixing quickly those 25 `NullSafeMutableLiveData` Lint errors and suppressing 10 of those, without even a failure.

Reply:
@ParaskP7 Yes, it's a CI failure, it didn't crash for me locally either. However, my local runs don't catch all the errors/failures, so I don't trust it all that much 😓 Here is the [initial failure](https://buildkite.com/automattic/wordpress-android/builds/4816#c2794969-a9e1-4384-8f27-20b5a3a9444e) which I thought was clear from [this comment](https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136331951), but I should have still clarified - sorry about that!

If you can branch from this PR and get a successful run in CI, I'd be happy to merge it into this PR. I really wanted to get it to work as well, but since it's a crash in a dependency, I couldn't come up with any ideas about how we can fix it in the project without updating the dependency.

I wonder if we could fix the "lint error" in the mentioned classes 🤔 There is no lint error reported due to the crash, but from the lint check type, we might be able to figure out how to improve the code which could fix the crash as well. Admittedly I didn't think about that yesterday.

Reply:
> @ParaskP7 Yes, it's a CI failure, it didn't crash for me locally either. However, my local runs don't catch all the errors/failures, so I don't trust it all that much 😓 Here is the [initial failure](https://buildkite.com/automattic/wordpress-android/builds/4816#c2794969-a9e1-4384-8f27-20b5a3a9444e) which I thought was clear from https://github.com/wordpress-mobile/WordPress-Android/pull/16632#issuecomment-1136331951, but I should have still clarified - sorry about that!

Thanks for sharing @oguzkocer , yes this [initial failure](https://buildkite.com/automattic/wordpress-android/builds/4816#c2794969-a9e1-4384-8f27-20b5a3a9444e) you shared was the same I was getting locally, lint failed with 37 errors, locally I was getting 34 errors but also 3 generic lint errors (total 38), but those [3 generic "LintError: Lint Failure" ones](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/0180f760-69c5-4714-970d-43ee98570e9f/c2794969-a9e1-4384-8f27-20b5a3a9444e/WordPress/build/reports/lint-results-wordpressVanillaRelease.html?response-content-type=text%2Fhtml&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIAQPCP3C7L2QXFF6U5%2F20220526%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220526T073835Z&X-Amz-Expires=600&X-Amz-Security-Token=IQoJb3JpZ2luX2VjELX%2F%2F%2F%2F%2F%2F%2F%2F%2F%2FwEaCXVzLWVhc3QtMSJGMEQCIDgOV1Jqxh2HRv9J9PYqgpzSmw0AzZ362SkRuivj8dI9AiAmoWBW0fL%2BJ1Wdgn%2Fa0PnDh9pZPiMN%2B8eQkDecLRfarSrbBAie%2F%2F%2F%2F%2F%2F%2F%2F%2F%2F8BEAAaDDAzMjM3OTcwNTMwMyIMBx3Qyhj2p8YC7IC5Kq8EXg63zVy99K1C44Z7ooYfppr7IPl80MKcEkecX3TFO6RdIgkITHccaL4L75pSIXqhhnEdVv5g6mP2MtODZRy9Zs0j%2F4pqrCfdg%2FlU1oSoRQCNhOKknZ%2FtZcHOwrCXThghHnst%2FKI%2F1WtAv60PI1%2BsACTF5IcRMZP7HXAiJD7U6AdCEaAN2Bmw%2BVCOkQP3oFB%2BqpAxpC%2BpjkLQuWbgKOmNbl1H7hbKeqF4lsgMyQ6pTFl5B7Wcto0aepBTxXFamj3xOp3EPu%2B5eTftNA0V2g%2FlDT0vXlBt3pfabDTLDaWbSLGwqKyNGbYaWwYseh%2BeSdGn0lM7F%2Bx1gphWdM%2B7N6SWqmlMOatFKunn4PBXQcE1quukYJSvG5n6iezuPyE0e7Xp8fmBdcZ9nQtZTybYc88myxxxk%2FO0iiMfEzCtdFvWsCLludRTKVCmal4HkTD%2FZuIIC4tvTMUjvyid6EdWuvBLjobBnc%2FEFyN80SKE8nUP57tEgWsIiXLJKVg9Tt3TbLdefofvB5kn9lIEDbDdWOBMjyCc02E95TpbjUybhb6qj2ToNvPnXsGe%2BJ0CJwpzEOcZiFBNYJBWp%2FerqbApAcwSDiyOUGG8onvEG9wfn8P9Py%2FnOb84tWIZ6GK0eQVA%2Bx13qkw1fxpMzuThtlVKAj2rCAtIzkuCXKLfOd3328HKDzS7Z0%2F2y1YgNo195W%2FI%2F5NohQIESxsjgRTd7dnYWYNmuJWLhufDBwIE0WXjIOFgyzDriryUBjqqAQjobB%2FM%2Bv1y7HpsdMOh4otcCa8tkZ0xQVSR23qhmpOSASv8xJ47NgDPBcvAUO72%2Flq4QxGhx1QMOgjVvyAqnDv4j15gbo2INNtA0xCgAx98T%2BNQ6tbX9lYkpmZ3kflj0GHP2o9EXC%2Bjjq4zN%2FqJs8o1Cfkl8kBuuuEWY5cvi869gd0m1dyAUITs8tI7015vqEizLB5%2BEKrPrSN1IQ5ffCo7SXExyQUlW1Gc&X-Amz-SignedHeaders=host&X-Amz-Signature=d7cf03937e509e8abdb7b48537192aa9af4c6b86d6a36e0be928a926b5e41b5e#LintError) all referred to specific classes within those other more specific 32 `NullSafeMutableLiveData` errors and as such it was evident for me that if those specific errors are fixed, then the generic ones would be probably fixed as well, and luckily enough they were fixed indeed (even though those 2 out of 3 `DiffCallback` generic lint errors were weird ones).

What throw me off was the fact that you mention that the lint is crashing due to the bugs you mentioned. To me, lint crashing vs. lint failing with errors mean separate things and thus I wanted to clarify that. Initial I though that lint couldn't even complete on your side and as the lint task was crashing. But, now I understand what you meant by that. 👍 

> If you can branch from this PR and get a successful run in CI, I'd be happy to merge it into this PR. I really wanted to get it to work as well, but since it's a crash in a dependency, I couldn't come up with any ideas about how we can fix it in the project without updating the dependency.

Yes, let me do that so that we could avoid disabling the `NullSafeMutableLiveData` error altogether for the whole project, which in most case is a good indication of something wrong going on and it would be great if engineers would fix that instead instead of ignoring it, on the spot in their PRs, so that we could avoid merging these issues into our `trunk`.

Please expect a PR from my side today. 👍 

> I wonder if we could fix the "lint error" in the mentioned classes 🤔 There is no lint error reported due to the crash, but from the lint check type, we might be able to figure out how to improve the code which could fix the crash as well. Admittedly I didn't think about that yesterday.

👍 💯

Reply:
> Please expect a PR from my side today. 👍

FYI: The PR is now ready: #16647

Reply:
@ParaskP7 It looks like the link task from https://github.com/wordpress-mobile/WordPress-Android/pull/16647 is also crashing.

> To me, lint crashing vs. lint failing with errors mean separate things and thus I wanted to clarify that.

They are objectively separate things and the issue we have is a crash, not a lint failure. This is clear from the lint report from the original commit as well as the new PR you've opened:

```
[../../src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/018100af-fba3-400f-b620-e682a3f7f814/018100b0-1b18-4561-83e9-e9632eb4668a/WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/sections/BaseStatsUseCase.kt): Unexpected failure during lint analysis of BaseStatsUseCase.kt (this is a bug in lint or one of the libraries it depends on)

Stack: ArrayIndexOutOfBoundsException:

You can run with --stacktrace or set environment variable LINT_PRINT_STACKTRACE=true to dump a full stacktrace to stdout.
[../../src/main/java/org/wordpress/android/ui/layoutpicker/LayoutCategoryDiffCallback.kt](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/018100af-fba3-400f-b620-e682a3f7f814/018100b0-1b18-4561-83e9-e9632eb4668a/WordPress/src/main/java/org/wordpress/android/ui/layoutpicker/LayoutCategoryDiffCallback.kt): Unexpected failure during lint analysis of LayoutCategoryDiffCallback.kt (this is a bug in lint or one of the libraries it depends on)

Message: Index 0 out of bounds for length 0

The crash seems to involve the detector androidx.lifecycle.lint.NonNullableMutableLiveDataDetector.
You can try disabling it with something like this:
    android {
        lintOptions {
            disable "NullSafeMutableLiveData"
        }
    }

Stack: ArrayIndexOutOfBoundsException:NonNullableMutableLiveDataDetector$createUastHandler$1.getFieldTypeReference(NonNullableMutableLiveDataDetector.kt:101)
←NonNullableMutableLiveDataDetector$createUastHandler$1.visitClass(NonNullableMutableLiveDataDetector.kt:88)
←UElementVisitor$DispatchPsiVisitor.visitClass(UElementVisitor.kt:600)
←UElementVisitor$DelegatingPsiVisitor.visitClass(UElementVisitor.kt:1208)
←AbstractKotlinUClass.accept(AbstractKotlinUClass.kt:42)
←ImplementationUtilsKt.acceptList(implementationUtils.kt:23)
←AbstractKotlinUClass.accept(AbstractKotlinUClass.kt:45)
←ImplementationUtilsKt.acceptList(implementationUtils.kt:23)

You can run with --stacktrace or set environment variable LINT_PRINT_STACKTRACE=true to dump a full stacktrace to stdout.
[../../src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsBlockDiffCallback.kt](https://s3.amazonaws.com/buildkiteartifacts.com/b26f1746-2f0c-49cc-bbd1-a45a5e24c351/e9cd0950-28ae-4b22-8385-c4c19564f99b/018100af-fba3-400f-b620-e682a3f7f814/018100b0-1b18-4561-83e9-e9632eb4668a/WordPress/src/main/java/org/wordpress/android/ui/stats/refresh/lists/StatsBlockDiffCallback.kt): Unexpected failure during lint analysis of StatsBlockDiffCallback.kt (this is a bug in lint or one of the libraries it depends on)

Stack: ArrayIndexOutOfBoundsException:

You can run with --stacktrace or set environment variable LINT_PRINT_STACKTRACE=true to dump a full stacktrace to stdout.
```

Reply:
Oh shoot, this is so very weird, I'll try and investigate further to figure this out, it does run successfully for me locally. Do this branch fails on you locally @oguzkocer ? 🤔

Reply:
@oguzkocer FYI, I think I managed to put this `NullSafeMutableLiveData` Lint error together to understand why it got introduced in the first place as part of this PR and how it might potentially be solved as part of [this accompanying](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) PR I just did.

My understanding below:
- As part of this PR, the transitive `lifecycle` dependency got update to `2.3.1` (from our current `2.2.0`).
- As part of this `2.3.1` lifecycle update and its [initial alpha01](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.3.0-alpha01) version Google introduced a new Lint check that warns you when setting a `null` value on ` MutableLiveData` that has been defined in Kotlin as non-null (see `New Features`).
- However, this newly added `NullSafeMutableLiveData` Lint check is buggy.
- First bug fix attempt with the [2.3.0-alpha06](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.3.0-alpha01) version.
- Second bug fix attempt with the [2.3.0-alpha07](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.3.0-alpha07) version.
- Last bug fix attempt with the [2.4.0-alpha02](https://developer.android.com/jetpack/androidx/releases/lifecycle#2.4.0-alpha02) version.
- As such, this made me think that we would need to update our `lifecycle` dependency to at least the `2.4.0` in an attempt to see if these `NullSafeMutableLiveData` Lint bugs will get resolved.
- The current latest stable `lifecycle` version is the `2.4.1`. It might be better to attempt and update to it instead.
- ❌ I just tried updating `lifecycle` to `2.4.1` with my [latest commit](https://github.com/wordpress-mobile/WordPress-Android/pull/16647/commits/7ee2401bc0a4f6197e314ee153d8ec62e31ab42e) on [my accompanying](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) PR, but it actually didn't help. 😞 
- ⚙️ I'll need to investigate further. Especially the fact that locally Lint succeeds on my side.

Reply:
Omg, it is getting trickier and trickier, I just noticed that we have a duplication of the `lifecycleVersion` version declaration in our root level `build.gradle` file and thus my update above failed as it did nothing, I updated the first declaration of `lifecycleVersion` to `2.4.1` but it got overwritten by the second one, which was again `2.2.0`. 🤷 

I am on it... stay tuned! 😅

Reply:
UPDATE: Ok, I am now making progress and updating to `2.4.1` works. However, it added 10 more `NullSafeMutableLiveData`, which to be honest, from the release notes, it is actually expected. This probably why Lint was failing with those weird errors that got it crashing. 🤞 

PS: I am running out of time today, I'll continue with this tomorrow. I converted the [my accompanying](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) PR back to draft for now.

Reply:
> UPDATE: Ok, I am now making progress and updating to 2.4.1 works. However, it added 10 more NullSafeMutableLiveData, which to be honest, from the release notes, it is actually expected. This probably why Lint was failing with those weird errors that got it crashing. 🤞

Just wanted to note that I saw this comment after I posted my last comment. 🤞 it works!

Reply:
UPDATE: And success, updating `lifecycle` to `2.4.1` and resolving/suppressing those 10 new `NullSafeMutableLiveData` lint errors did the trick. As such, the [accompanying PR](https://github.com/wordpress-mobile/WordPress-Android/pull/16647) is now ready for review again. 🎉 

Cc @oguzkocer @ashiagr